src/evox/problems/numerical/lsmop.py

```diff
@@ -0,0 +1,755 @@
+        return jnp.sum(x**2, axis=1, keepdims=True)
+
+    @staticmethod
+    def _Giewank(x):
+        f = (
+            jnp.sum(x**2, axis=1, keepdims=True) / 4000
+            - jnp.prod(
+                jnp.cos(
+                    x
+                    / jnp.tile(
+                        jnp.sqrt(jnp.arange(1, jnp.shape(x)[1] + 1)),
+                        (jnp.shape(x)[0], 1),
+                    )
+                ),
+                axis=1,
+                keepdims=True,
+            )
+            + 1
+        )
+        return f
+
+    @staticmethod
+    def _Schwefel(x):
+        return jnp.max(jnp.abs(x), keepdims=True, axis=1)
+
+    @staticmethod
+    def _Rastrigin(x):
+        f = jnp.sum(x**2 - 10 * jnp.cos(2 * jnp.pi * x) + 10, axis=1, keepdims=True)
+        return f
+
+    @staticmethod
+    def _Rosenbrock(x):
+        f = jnp.sum(
+            100 * ((x[:, : jnp.shape(x)[1] - 1]) ** 2 - x[:, 1 : jnp.shape(x)[1]]) ** 2
+            + (x[:, : jnp.shape(x)[1] - 1] - 1) ** 2,
+            axis=1,
+            keepdims=True,
+        )
+        return f
+
+    @staticmethod
+    def _Ackley(x):
+        # np.seterr(divide='ignore', invalid='ignore')
+        f = (
+            20
+            - 20
+            * jnp.exp(
+                -0.2
+                * jnp.sqrt(jnp.sum(x**2, axis=1, keepdims=True) / jnp.shape(x)[1])
+            )
+            - jnp.exp(
+                jnp.sum(jnp.cos(2 * jnp.pi * x), axis=1, keepdims=True)
+                / jnp.shape(x)[1]
+            )
+            + jnp.exp(1)
+        )
+        if jnp.isnan(f[0]):
+            f = jnp.zeros([jnp.shape(f)[0], jnp.shape(f)[1]])
+        return f
+
+    @staticmethod
+    def _Griewank(x):
+        f = (
+            jnp.sum(x**2, axis=1, keepdims=True) / 4000
+            - jnp.prod(
+                jnp.cos(
+                    x
+                    / jnp.tile(
+                        jnp.sqrt(jnp.arange(1, jnp.shape(x)[1] + 1)),
+                        (jnp.shape(x)[0], 1),
+                    )
+                ),
+                axis=1,
+                keepdims=True,
+            )
+            + 1
+        )
+        return f
+
+
+class LSMOP1(LSMOP):
+    def __init__(self, d=None, m=None, ref_num=1000):
+        super().__init__(d, m, ref_num)
+        if m is None:
+            self.m = 3
+        else:
+            self.m = m
+        if d is None:
+            self.d = self.m + 4
+        else:
+            self.d = d
+
+    def evaluate(self, state: chex.PyTreeDef, X: chex.Array):
+        m = self.m
+        n, d = jnp.shape(X)
+        # X[:, m - 1: d] = (1 + jnp.tile(jnp.arange(m, d + 1) / d, (n, 1))) * X[:, m - 1:d] - jnp.tile(X[:, :1] * 10, (1, d - m + 1))
+        X = X.at[:, m - 1 : d].set(
+            (1 + jnp.tile(jnp.arange(m, d + 1) / d, (n, 1))) * X[:, m - 1 : d]
+            - jnp.tile(X[:, :1] * 10, (1, d - m + 1))
+        )
+        g = jnp.zeros([n, m])
+        for i in range(0, m, 2):
```

Comment:
Usually, you want to avoid for loop, especially for-range loop, it's usually bad to use for loop in functional programming.
If you really need it, please add comment to explain this nested for loop, otherwise no one can understand.

Reply:
> Usually, you want to avoid for loop, especially for-range loop, it's usually bad to use for loop in functional programming. If you really need it, please add comment to explain this nested for loop, otherwise no one can understand.

Maybe it's partially correct. However, in my opinion, the no.1 importance of a code is its readability. If you insist functional programming, then maybe the class definitions themselves shall be replaced by data descriptions as well since they are not functional :( , which seems unnecessary. In this case, the simple for-range loop provides OK readability and replacing them with `lax.fori_loop` not only damages the readability but also reduces runtime efficiency potentially. Therefore, it is better to completely annihilate loops (if possible) in pull request(s) for enhancement(s) rather than adding new feature(s).